homeassistant/components/fan/xiaomi_miio.py

```diff
@@ -0,0 +1,354 @@
+ATTR_LEVEL = 'level'
+
+SUCCESS = ['ok']
+
+SERVICE_SET_BUZZER_ON = 'set_buzzer_on'
+SERVICE_SET_BUZZER_OFF = 'set_buzzer_off'
+SERVICE_SET_LED_ON = 'set_led_on'
+SERVICE_SET_LED_OFF = 'set_led_off'
+SERVICE_SET_FAVORITE_LEVEL = 'set_favorite_level'
+SERVICE_SET_LED_BRIGHTNESS = 'set_led_brightness'
+
+AIRPURIFIER_SERVICE_SCHEMA = vol.Schema({
+    vol.Optional(ATTR_ENTITY_ID): cv.entity_ids,
+})
+
+SERVICE_SCHEMA_LED_BRIGHTNESS = AIRPURIFIER_SERVICE_SCHEMA.extend({
+    vol.Required(ATTR_BRIGHTNESS):
+        vol.All(vol.Coerce(int), vol.Clamp(min=0, max=2))
+})
+
+SERVICE_SCHEMA_FAVORITE_LEVEL = AIRPURIFIER_SERVICE_SCHEMA.extend({
+    vol.Required(ATTR_LEVEL):
+        vol.All(vol.Coerce(int), vol.Clamp(min=0, max=17))
+})
+
+SERVICE_TO_METHOD = {
+    SERVICE_SET_BUZZER_ON: {'method': 'async_set_buzzer_on'},
+    SERVICE_SET_BUZZER_OFF: {'method': 'async_set_buzzer_off'},
+    SERVICE_SET_LED_ON: {'method': 'async_set_led_on'},
+    SERVICE_SET_LED_OFF: {'method': 'async_set_led_off'},
+    SERVICE_SET_FAVORITE_LEVEL: {
+        'method': 'async_set_favorite_level',
+        'schema': SERVICE_SCHEMA_FAVORITE_LEVEL},
+    SERVICE_SET_LED_BRIGHTNESS: {
+        'method': 'async_set_led_brightness',
+        'schema': SERVICE_SCHEMA_LED_BRIGHTNESS},
+}
+
+
+# pylint: disable=unused-argument
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the air purifier from config."""
+    from mirobo import AirPurifier, DeviceException
+    if PLATFORM not in hass.data:
+        hass.data[PLATFORM] = {}
+
+    host = config.get(CONF_HOST)
+    name = config.get(CONF_NAME)
+    token = config.get(CONF_TOKEN)
+
+    _LOGGER.info("Initializing with host %s (token %s...)", host, token[:5])
+
+    try:
+        air_purifier = AirPurifier(host, token)
+
+        xiaomi_air_purifier = XiaomiAirPurifier(name, air_purifier)
+        hass.data[PLATFORM][host] = xiaomi_air_purifier
+    except DeviceException:
+        raise PlatformNotReady
+
+    async_add_devices([xiaomi_air_purifier], update_before_add=True)
+
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Map services to methods on XiaomiAirPurifier."""
+        method = SERVICE_TO_METHOD.get(service.service)
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+        entity_ids = service.data.get(ATTR_ENTITY_ID)
+        if entity_ids:
+            target_air_purifiers = [air for air in hass.data[PLATFORM].values()
+                                    if air.entity_id in entity_ids]
+        else:
+            target_air_purifiers = hass.data[PLATFORM].values()
+
+        update_tasks = []
+        for air_purifier in target_air_purifiers:
+            yield from getattr(air_purifier, method['method'])(**params)
+
+        for air_purifier in target_air_purifiers:
+            update_tasks.append(air_purifier.async_update_ha_state(True))
```

Comment:
Could/should this be merged with the above for loop, as it loops through the same list?